File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        
        int left = 0;
        int right = n - 1;
        int leftMax = 0;
        int rightMax = 0;
        
        int ans = 0;
        
        while(left <= right) {
            if(height[left] <= height[right]) {
                if(height[left] >= leftMax) {
                    leftMax = height[left];
                } else {
                    ans += leftMax - height[left];
                }
                left++;
            } else {
                if(height[right] >= rightMax) {
                    rightMax = height[right];
                } else {
                    ans += rightMax - height[right];
                }
                right--;
            }
        }
        
        return ans;
    }
}

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        
        vector<int> leftMax(n, 0);
        int lm = 0;
        for(int i = 0; i < n; i++) {
            leftMax[i] = lm;
            lm = max(lm, height[i]);
        }
        
        vector<int> rightMax(n, 0);
        int rm = 0;
        for(int i = n - 1; i >= 0; i--) {
            rightMax[i] = rm;
            rm = max(rm, height[i]);
        }
        
        int ans = 0;
        for(int i = 0; i < n; i++) {
            int minHt = min(leftMax[i], rightMax[i]);
            if(minHt > height[i]) {
                ans += minHt - height[i];
            }
        }
        
        return ans;
    }
};